### PR TITLE
Add onlyOwner modifier to token handler facets

### PR DESCRIPTION
**Pull request by TJ-Everett**

### CODE REVIEW

**Review by mpetersoCode55 — APPROVED** [review]
Good work Tayler! Looks great. Approved for me

File: src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol [hunk 1]
```diff
@@ -30,67 +32,129 @@ contract ERC721NonTaggedRuleFacet is
...
      */
-    function checkNonTaggedRules(ActionTypes action, address _from, address _to, uint256 _amount, uint256 _tokenId) external {
+    function checkNonTaggedRules(ActionTypes action, address _from, address _to, uint256 _amount, uint256 _tokenId) external onlyOwner {
         _from;
         HandlerBaseS storage handlerBaseStorage = lib.handlerBaseStorage();
-        mapping(ActionTypes => Rule[]) storage accountApproveDenyOracle = lib.accountApproveDenyOracleStorage().accountApproveDenyOracle;
-        /// The action type determines if the _to or _from is checked by the oracle
-        /// _from address is checked for Burn and Sell action types  
-        if (action == ActionTypes.BURN || action == ActionTypes.SELL){
-            IRuleProcessor(handlerBaseStorage.ruleProcessor).checkAccountApproveDenyOracles(accountApproveDenyOracle[action], _from);
```
Comment (ShaneDuncan602) [inline src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:40]: Why are we getting rid of these checks?

Comment (ShaneDuncan602) [reply src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:40]: Nevermind, I see you put them in the internal function. It's a lot cleaner now.

File: src/client/token/handler/diamond/TradingRuleFacet.sol [hunk 2]
```diff
@@ -22,54 +24,74 @@ contract TradingRuleFacet is HandlerAccountMaxTradeSize, HandlerTokenMaxBuySellV
...
-                    IERC20Decimals(msg.sender).totalSupply(),  
-                    _amount,  
-                    maxVolume.lastPurchaseTime,  
-                    maxVolume.boughtInPeriod
-                );
-                maxVolume.lastPurchaseTime = uint64(block.timestamp); /// update with new blockTime if rule check is successful
-            }
+            _checkTradeRulesBuyAction(_to, toTags, _amount, action);
+        } else if(action == ActionTypes.SELL) {
+            _checkTradeRulesSellAction(_from, fromTags, _amount, action);
```
Comment (ShaneDuncan602) [inline src/client/token/handler/diamond/TradingRuleFacet.sol:31]: This is so much better!

**Comment by TJ-Everett** [comment]
The crux of the issue is that some facets are missing the necessary access controls to prevent an attacker from manipulating the state of the diamond. For example, the ERC721NonTaggedRuleFacet.checkNonTaggedRules function is an external function that updates diamond state. Thus, an attacker can directly invoke this function through the HandlerDiamond contract and update the state of the system. We found this issue in the following facets:
[ERC721NonTaggedRuleFacet.checkNonTaggedRules](https://github.com/thrackle-io/tron/blob/01302c38891cf1de78690a6475325d1f55570270/src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol#L33-L33)
[TradingRuleFacet.checkTradingRules](https://github.com/thrackle-io/tron/blob/01302c38891cf1de78690a6475325d1f55570270/src/client/token/handler/diamond/TradingRuleFacet.sol#L25-L25)
[ERC20NonTaggedRuleFacet.checkNonTaggedRules](https://github.com/thrackle-io/tron/blob/01302c38891cf1de78690a6475325d1f55570270/src/client/token/handler/diamond/ERC20NonTaggedRuleFacet.sol#L20-L20)

The easiest remediation is to create an external function that has an onlyOwner modifier which passes the arguments to an internal function that handles the core logic.

File: src/client/token/handler/diamond/StorageLib.sol [hunk 3]
```diff
@@ -36,6 +41,61 @@ library StorageLib {
...
+    /**
+     * @dev Function to store the Initialized flag for Trading Rule Facet 
+     * @return ds Data Storage of the Initialized flag
+     */
+    function tradingRulesInitializedStorage() internal pure returns (InitializedS storage ds) {
+        bytes32 position = TRADING_RULES_FACET_INITIALIZED_POSITION;
+        assembly {
+            ds.slot := position
+        }
+    }
```
Comment (oscarsernarosero) [inline src/client/token/handler/diamond/StorageLib.sol]: Not sure what these positions are storing

File: src/client/token/handler/diamond/TradingRuleFacet.sol (hunk 2 above)
Comment (oscarsernarosero) [reply src/client/token/handler/diamond/TradingRuleFacet.sol:31]: Agree. Great job!

File: test/client/token/ERC20/integration/ERC20CommonTests.t.sol [hunk 4]
```diff
@@ -2529,4 +2529,24 @@ abstract contract ERC20CommonTests is TestCommonFoundry, DummyAMM, ERC20Util {
...
+        erc20NonTaggedTestFacet.checkNonTaggedRules(user1, user, 10, ActionTypes.P2P_TRANSFER);
+        vm.expectRevert("UNAUTHORIZED");
+        tradeRuleTestFacet.checkTradingRules(user1, user, createBytes32Array("Oscar"), createBytes32Array("Oscar"), 10, ActionTypes.P2P_TRANSFER);
+
+        /// test that users cannot call the facets directly through the proxy address 
+        vm.expectRevert("UNAUTHORIZED");
+        ERC20TaggedRuleFacet(address(applicationCoinHandler)).checkTaggedAndTradingRules(10, 10, user1, user, 10, ActionTypes.P2P_TRANSFER);
+        vm.expectRevert("UNAUTHORIZED");
+        ERC20NonTaggedRuleFacet(address(applicationCoinHandler)).checkNonTaggedRules(user1, user, 10, ActionTypes.P2P_TRANSFER);
+    }
```
Comment (oscarsernarosero) [inline test/client/token/ERC20/integration/ERC20CommonTests.t.sol]: I think it's better if we split this test into multiple tests. One per facet.

File: test/client/token/ERC721/integration/ERC721CommonTests.t.sol [hunk 5]
```diff
@@ -2959,4 +2959,21 @@ abstract contract ERC721CommonTests is TestCommonFoundry, ERC721Util {
...
+        vm.expectRevert("UNAUTHORIZED");
+        erc721TaggedTestFacet.checkTaggedAndTradingRules(10, 10, user1, user, 10, ActionTypes.P2P_TRANSFER);
+        vm.expectRevert("UNAUTHORIZED");
+        erc721NonTaggedTestFacet.checkNonTaggedRules(ActionTypes.P2P_TRANSFER, user1, user, 10, 0);
+
+        /// test that users cannot call the facets directly through the proxy address 
+        vm.expectRevert("UNAUTHORIZED");
+        ERC721TaggedRuleFacet(address(applicationNFTHandler)).checkTaggedAndTradingRules(10, 10, user1, user, 10, ActionTypes.P2P_TRANSFER);
+        vm.expectRevert("UNAUTHORIZED");
+        ERC721NonTaggedRuleFacet(address(applicationNFTHandler)).checkNonTaggedRules(ActionTypes.P2P_TRANSFER, user1, user, 10, 0);
```
Comment (oscarsernarosero) [inline test/client/token/ERC721/integration/ERC721CommonTests.t.sol]: same here

File: test/client/token/ERC20/integration/ERC20CommonTests.t.sol (hunk 4 above)
Comment (oscarsernarosero) [reply test/client/token/ERC20/integration/ERC20CommonTests.t.sol]: Also, it looks like you skipped the applicationCoinHandler-style call on the trading rules.

**Review by oscarsernarosero — CHANGES_REQUESTED** [review]
Great job here Tayler. I have a doubt and a couple comments on the tests although nothing major.

File: src/client/token/handler/diamond/StorageLib.sol (hunk 3 above)
Comment (TJ-Everett) [reply src/client/token/handler/diamond/StorageLib.sol]: Thank you for the catch! I initially thought we would need initializer functions for ownership modifiers before I dug deep enough into the imports. Then didn't go back and remove. They are gone now.

File: test/client/token/ERC20/integration/ERC20CommonTests.t.sol (hunk 4 above)
Comment (TJ-Everett) [reply test/client/token/ERC20/integration/ERC20CommonTests.t.sol]: Great idea splitting these out!

**Review by oscarsernarosero — APPROVED** [review]
Looks great Tayler!

**Pull request merged**